File: src/main/java/nz/co/searchwellington/controllers/ajax/TagAjaxController.java
package nz.co.searchwellington.controllers.ajax;

import java.io.IOException;
import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import nz.co.searchwellington.repositories.ContentRetrievalService;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;

import uk.co.eelpieconsulting.common.views.ViewFactory;

@Controller
public class TagAjaxController extends BaseAjaxController {
	
    private static Logger log = Logger.getLogger(TagAjaxController.class);

    private ContentRetrievalService contentRetrievalService;
    
    @Autowired
    public TagAjaxController(ViewFactory viewFactory, ContentRetrievalService contentRetrievalService) {
    	this.viewFactory = viewFactory;
		this.contentRetrievalService = contentRetrievalService;
	}
    
	@Override
	@RequestMapping("/ajax/tags")
	public ModelAndView handleRequest(HttpServletRequest request, HttpServletResponse response) throws IOException {
		return super.handleRequest(request, response);
	}
	
	protected List<String> getSuggestions(String q) {
        log.debug("Looking up possible tags starting with: " + q);
        return contentRetrievalService.getTagNamesStartingWith(q);
	}
        
}

File: src/main/scala/nz/co/searchwellington/controllers/ajax/TagAjaxController.scala
package nz.co.searchwellington.controllers.ajax

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.LoggedInUserFilter
import nz.co.searchwellington.model.User
import nz.co.searchwellington.repositories.ContentRetrievalService
import org.apache.log4j.Logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.servlet.ModelAndView
import uk.co.eelpieconsulting.common.views.ViewFactory

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import scala.concurrent.Future

@Controller class TagAjaxController @Autowired()(val loggedInUserFilter: LoggedInUserFilter,
                                                 val viewFactory: ViewFactory,
                                                 contentRetrievalService: ContentRetrievalService) extends BaseAjaxController with ReasonableWaits {

  private val log = Logger.getLogger(classOf[TagAjaxController])

  @GetMapping(Array("/ajax/tags"))
  override def handleRequest(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    super.handleRequest(request, response)
  }

  override protected def getSuggestions(q: String, loggedInUser: Option[User]): Future[Seq[String]] = {
    log.debug("Looking up possible tags starting with: " + q)
    contentRetrievalService.getTagNamesStartingWith(q, loggedInUser)
  }

}
